### PR TITLE
Readme updates for root and the capi

### PR DESCRIPTION
**Pull request by pittma**

### CODE REVIEW

File: modality-probe-capi/README.md [hunk 1]
```diff
@@ -1,53 +1,172 @@
...
+
+```c
+#define DEFAULT_PROBE_SIZE = 7000;
+modality_probe * probe_g = MODALITY_PROBE_NULL_INITIALIZER;
+
+void twist(double x, double y, double z) {
+    int result = MODALITY_PROBE_RECORD(
+        probe_g,
+        TWISTED,
+        "tags=actuation",
```
Comment (jonlamb-gh) [inline modality-probe-capi/README.md:70]: This is using the old tagging, now we have `MODALITY_TAGS(...)`: `MODALITY_TAGS("actuation")`

File: modality-probe-capi/README.md [hunk 2]
```diff
@@ -1,53 +1,172 @@
...
+}
+
+int main() {
+    uint8_t * destination = (uint8_t*)malloc(DEFAULT_PROBE_SIZE);
+    modality_probe_error result = modality_probe_initialize(
+        destination,
+        DEFAULT_PROBE_SIZE,
+        CONTROLLER,
+        probe_g
+    );
```
Comment (jonlamb-gh) [inline modality-probe-capi/README.md:83]: Should be using the cli-friendly macro here:
```c
result = MODALITY_PROBE_INIT(
        destination,
        DEFAULT_PROBE_SIZE,
        CONTROLLER,
        NULL,    /* No restart tracking */
        NULL,
        &probe_g
        MODALITY_TAGS("controller"),
        "The controller");
```

File: modality-probe-capi/README.md [hunk 3]
```diff
@@ -1,53 +1,172 @@
...
+your existing logging infrastructure by embedding a probe’s current
+clock value into your log messages:
+
+```c
+#ifdef MODALITY_PROBE_MACROS_ENABLED
+    modality_probe_instant now = modality_probe_now(CONTROLLER);
+    LOG(
+        “controller: sent a twist command, probe clock (%d, %d)”,
+        now.clock.id,
+        now.clock.count
```
Comment (jonlamb-gh) [inline modality-probe-capi/README.md:126]: The new `modality_probe_instant` type has changed a bit, probably ought to be (`now.clock.id`, `now.clock.epoch`, `now.clock.ticks`, `now.event_count`).

File: README.md [hunk 4]
```diff
@@ -1,312 +1,157 @@
-# ekotrace
+# `modality-probe`
 
-A distributed causal tracing system suitable for embedded use.
+`modality-probe` is an open source part of Auxon’s Modality™
+continuous verification & validation platform. Its role is to record
+events and track their causal relationships between the different
+parts of your system. Why? Because being able to stitch together a
+causal history of a system, particularly a system _under test_,
+provides a high-resolution lens into _what happened_. This history can
```
[the remark this answers is not in the document]
Comment (pittma) [reply README.md:8]: My intention here was to a little dramatic; "testing" is repeated in the next sentence, saying it here I think interrupts the flow of this paragraph.

File: README.md [hunk 5]
```diff
@@ -1,312 +1,157 @@
...
-
-```rust
-// Initialization of an ekotrace instance only needs to happen once
-// in the lifetime of a thread / execution context
-let ekotrace_foo = Ekotrace::initialize_at(&mut storage_bytes, LOCATION_ID_FOO)?;
-ekotrace_foo.record_event(EVENT_A);
-ekotrace_foo.record_event(EVENT_B);
+```shell
+modality-probe $ cd package/debian
+modality-probe/packing/debian $ ./build.sh
```
[the remark this answers is not in the document]
Comment (pittma) [reply README.md:43]: Well, right now, we don't deposit releases on merge or anything. I saw this as a shorter way to get to having everything in the right place for it to "just work". I'm fine to skip this altogether, though, with the intention to upload things to apt and/or make use of Github releases.

File: README.md [hunk 6]
```diff
@@ -1,312 +1,157 @@
...
+#### Other *nixes (*nices?)
 
-The `ekotrace` client provides a system-agnostic approach to reporting. `ekotrace` instances
-write their report data into a byte buffer, which can be sent to a report collecting backend
-whichever way the system under test prefers. Report content can be treated as opaque
-bytes by the embedded system.
+Cargo can use something called "workspaces" to build groups of
+libraries or applications that share a source tree. `modality-probe`
+uses workspaces, and we can ask cargo to build all of the members of a
+workspace with the `--all` switch:
```
[the remark this answers is not in the document]
Comment (pittma) [reply README.md:65]: I don't think it's bad to explain what all is doing here.

File: README.md [hunk 7]
```diff
@@ -1,312 +1,157 @@
...
-        .arg("--lang")
-        .arg("Rust")
-        .arg("events.csv")
-        .arg("tracers.csv")
-        .output()
-        .expect("ekotrace-header-gen failed unexpectedly");
-    io::stderr().write_all(&output.stderr).expect("Could not write to stderr");
-    events_src.write_all(&output.stdout).expect("Could not write to stdout");
-    assert!(output.status.success());
+    let mut storage = [0u8; 1024];
```
[the remark this answers is not in the document]
Comment (pittma) [reply README.md:133]: I _think_ it'd have to be a `static mut` which is not socially acceptable. Is that wrong? What say you @jonlamb-gh?

File: modality-probe-capi/README.md [hunk 8]
```diff
@@ -1,53 +1,172 @@
-# ekotrace-capi
+# `modality-probe-capi`
+## Overview
+`modality-probe-capi` is a linkable C library that one can use to
+record events into a probe’s log, share log snapshots across probes,
+and “report” a probe’s log to a collector. The library is also
+targeted by the [code and manifest generation that the CLI does](link
+to cli).
```
[the remark this answers is not in the document]
Comment (pittma) [reply modality-probe-capi/README.md:7]: Manifest generation, which I intend to link to here.

File: modality-probe-capi/README.md [hunk 9]
```diff
@@ -1,53 +1,172 @@
...
-```shell script
-cargo build --release
+### Cross-platform Builds
+
+In order to build this library for platforms other than your host
+system, we recommend the use of cargo-xbuild. The default usage path
+for obtaining a new target's toolchain and building for likely looks
+something like the following. For more detailed information about
+cross-compilation for Rust, see
+[xbuild](https://github.com/rust-osdev/cargo-xbuild) and
```
[the remark this answers is not in the document]
Comment (pittma) [reply modality-probe-capi/README.md:29]: Ah, I didn't know we'd moved on from `xbuild`. This was all cribbed and slightly updated from what @jonlamb-gh had previously written here.

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jonlamb-gh) [reply README.md:133]: I read the suggestion as "use a const for the storage size", something like `DEFAULT_PROBE_SIZE` used in the C example?

File: modality-probe-capi/README.md [hunk 10]
```diff
@@ -1,53 +1,172 @@
...
+[rustc](https://github.com/japaric/rust-cross).
+
+```shell
+modality-probe/modality-probe-capi $ rustup update
```
[the remark this answers is not in the document]
Comment (pittma) [reply modality-probe-capi/README.md:33]: According to the rust-toolchain file here, the C API requires nightly. What do you mean by putting it in our Cargo.toml?

IIUC, if they have nightly installed, this rustup update will update it to the latest, and if they don't, then when they run any build commands the latest nightly will be installed on account of the toolchain file. Do you agree?

File: README.md [hunk 11]
```diff
@@ -1,312 +1,157 @@
...
+cargo command available to you and the commands in the following
+section should work.
 
-Causal tracing works without relying on system clocks at all.
+### Build it
 
-`ekotrace` can trace one or many embedded components and provides
-information about the impact of each component on the system as
-a whole.
+#### Linux (Debian-flavored)
```
Comment (jonlamb-gh) [inline README.md:29]: We're about to remove Debian packaging in favor of tarballs in https://github.com/auxoncorp/modality-probe/pull/207.

Comment (jonlamb-gh) [reply README.md:29]: Maybe this is ok for now though, and we can update afterwards?

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (pittma) [reply README.md:133]: ah crap, yeah, that makes sense.

File: README.md [hunk 12]
```diff
@@ -1,312 +1,157 @@
...
-## Automated Workflow Example
-
-Start using the `ekotrace` API, don't worry about defining your event identifiers.
+Then you can use it to record events.
 
-```rust
-// src/tracing_ids.rs mod generated by ekotrace-header-gen
-use crate::tracing_ids::*;
+```c
+#define DEFAULT_PROBE_SIZE = 7000;
```
Comment (jonlamb-gh) [inline README.md:113]: 7,000 bytes is pretty big, can we make this 1024 to match the other snippets?

Comment (jonlamb-gh) [reply README.md:113]: Also, minor syntax fix: `#define DEFAULT_PROBE_SIZE (1024)`

File: README.md [hunk 13]
```diff
@@ -1,312 +1,157 @@
...
+modality_probe * probe_g = MODALITY_PROBE_NULL_INITIALIZER;
```
Comment (jonlamb-gh) [inline README.md:114]: Minor nit: usually globals are prefixed with `g_`, rather than suffixed.

File: README.md [hunk 14]
```diff
@@ -1,312 +1,157 @@
...
-
-// Record events
-match bar() {
-    true => ekotrace_foo.record_event(EVENT_A),
-    false => ekotrace_foo.record_event(EVENT_B),
+void twist(double x, double y, double z) {
+    int result = MODALITY_PROBE_RECORD(
+        probe_g,
+        TWISTED,
+        TAGS("actuation"),
```
Comment (jonlamb-gh) [inline README.md:120]: `s/TAGS/MODALITY_TAGS`

File: README.md [hunk 15]
```diff
@@ -1,312 +1,157 @@
...
-        .arg("Rust")
-        .arg("events.csv")
-        .arg("tracers.csv")
-        .output()
-        .expect("ekotrace-header-gen failed unexpectedly");
-    io::stderr().write_all(&output.stderr).expect("Could not write to stderr");
-    events_src.write_all(&output.stdout).expect("Could not write to stdout");
-    assert!(output.status.success());
+    let mut storage = [0u8; 1024];
+    let tracer = try_initialize_at!(&mut storage, LID_B, TAGS("actuation"), "Twister")
```
Comment (jonlamb-gh) [inline README.md:134]: `s/TAGS/tags!/`

File: modality-probe-capi/README.md [hunk 16]
```diff
@@ -1,53 +1,172 @@
...
+probe as a global called `probe_g`. You’ll also note that there are a
+few seemingly undefined symbols: `TWISTED` and `CONTROLLER`. These are
+given definitions by making use of Modality Probe’s CLI sub-command
+`header-gen` which you can read more about [here TODO(link)]. The
+short of it is that you’ll need to run the header-gen tool against
+your codebase before it can be compiled because Modality Probe
+actually generates the definitions to those symbols.
+
+```c
+#define DEFAULT_PROBE_SIZE = 7000;
```
[the remark this answers is not in the document]
Comment (jonlamb-gh) [reply modality-probe-capi/README.md:63]:
```c
#include <modality/probe.h>

#define DEFAULT_PROBE_SIZE (1024)
```

File: modality-probe-capi/README.md [hunk 17]
```diff
@@ -1,53 +1,172 @@
...
+agent must periodically “report” what’s in its log, i.e., the events
+and interactions it’s recorded. To do that, use the report API:
+
+```c
+uint8_t * log_storage = (uint8_t*)malloc(LOG_STORAGE_SIZE);
+size_t bytes_written;
+result = modality_probe_report(
+    probe_g,
+    log_storage,
+    lOG_STORAGE_SIZE,
```
Comment (jonlamb-gh) [inline modality-probe-capi/README.md:151]: `s/lOG/LOG`

File: README.md (hunk 15 above)
Comment (jonlamb-gh) [reply README.md:134]: Also missing `RestartCounterProvider::NoRestartTracking` initializer arg

File: README.md (hunk 11 above)
Comment (pittma) [reply README.md:29]: I think I agree with @mullr, and that this should just come out.

File: README.md [hunk 18]
```diff
@@ -1,312 +1,157 @@
...
 
-ekotrace manifest-gen --events-csv-file events.csv --tracers-csv-file tracers.csv ./
+#### In C
 
-ekotrace header-gen --lang Rust events.csv tracers.csv > tracing_ids/mod.rs
+You first need to initialize your tracer.
 
-cargo run --example event-recording
+```c
+int main() {
```
Comment (jonlamb-gh) [inline README.md:99]: Few C nits:

```c
int main(int argc, char **argv)
{
    ...
}
```
With the macro's:

```c 
result = MODALITY_PROBE_INIT(
        destination,
        DEFAULT_PROBE_SIZE,
        CONTROLLER,
        NULL,    /* No restart tracking */
        NULL,
        &g_probe,
        MODALITY_TAGS("controller"),
        "The controller");
```

**Pull request merged**